File: src/examples/go/generation/typesafe-native.go
// ax-example:start
// title: Go Jev Native Questions
// group: generation
// description: Uses structured criteria, native scoring, model discovery, and probability-based decisions.
// provider: typesafe
// env: TYPESAFE_APIKEY
// level: advanced
// order: 36
// ax-example:end
package main

import (
	"context"
	"encoding/json"
	"fmt"
	ax "github.com/ax-llm/ax/packages/go"
	"os"
	"time"
)

func main() {
	ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
	defer cancel()
	client := ax.Typesafe(map[string]ax.Value{"api_key": os.Getenv("TYPESAFE_APIKEY")})
	models, err := client.ListModels(ctx, nil)
	if err != nil {
		panic(err)
	}
	if len(models) == 0 {
		panic("Empty model catalog")
	}
	request := ax.TypesafeRequest{
		State: map[string]ax.Value{
			"ticket":  "Checkout is unavailable for all customers after the latest deployment.",
			"account": map[string]ax.Value{"tier": "enterprise", "notes": nil},
		},
		Questions: map[string]ax.TypesafeQuestion{
			"urgent": {Type: "noul",
				Instructions: map[string]ax.Value{"question": "Does this need immediate attention?"},
				Criteria:     map[string]ax.Value{"true": "Customers cannot complete a core task", "false": "Routine request"}},
			"team": {Type: "choice", Instructions: "Who should handle the ticket?",
				Criteria: map[string]ax.Value{"support": "Usage guidance",
					"billing": map[string]ax.Value{"scope": "Invoices and payments"}, "engineering": "Product failures"}},
			"severity": {Type: "score", Instructions: "Rate customer impact",
				Criteria: []ax.Value{"Minor inconvenience", "One task blocked", "Core task unavailable", "Widespread outage"}},
		},
	}
	response, err := client.SystemOne(ctx, request, nil)
	if err != nil {
		panic(err)
	}
	probability := response.Answers["urgent"].Noul
	score := response.Answers["severity"].Score
	if probability < 0 || probability > 1 || score < 0 || score > 3 {
		panic("Invalid answer bounds")
	}
	// Apply thresholds and custom score scales in application code.
	output := map[string]any{"page_on_call": probability >= 0.9, "severity_1_to_5": 1 + 4*score/3, "response": response}
	data, err := json.MarshalIndent(output, "", "  ")
	if err != nil {
		panic(err)
	}
	fmt.Println(string(data))
}

File: src/examples/cpp/generation/typesafe-native.cpp
// ax-example:start
// title: Cpp Jev Native Questions
// group: generation
// description: Uses structured criteria, native scoring, model discovery, and probability-based decisions.
// provider: typesafe
// env: TYPESAFE_APIKEY
// level: advanced
// order: 36
// ax-example:end
#include "axllm/axllm.hpp"
#include <cstdlib>
#include <iostream>
#include <stdexcept>
using namespace axllm;

static std::string key(const char* name) { const char* value = std::getenv(name); if (!value || !*value) throw std::runtime_error(std::string("Set ") + name); return value; }
int main() {
  auto client = typesafe(object({{"api_key", key("TYPESAFE_APIKEY")}}));
  if (client.list_models().empty()) throw std::runtime_error("Empty catalog");
  TypesafeRequest request;
  request.state = object({
    {"ticket", "Checkout is unavailable for all customers after the latest deployment."},
    {"account", object({{"tier", "enterprise"}, {"notes", Value()}})}
  });
  request.questions = {
    {"urgent", {"noul", object({{"question", "Does this need immediate attention?"}}),
      object({{"true", "Customers cannot complete a core task"}, {"false", "Routine request"}})}},
    {"team", {"choice", "Who should handle the ticket?",
      object({{"support", "Usage guidance"}, {"billing", object({{"scope", "Invoices and payments"}})},
              {"engineering", "Product failures"}})}},
    {"severity", {"score", "Rate customer impact",
      array({"Minor inconvenience", "One task blocked", "Core task unavailable", "Widespread outage"})}}
  };
  auto response = client.system_one(request);
  double probability = std::get<TypesafeNoul>(response.answers.at("urgent")).noul;
  double score = std::get<TypesafeScore>(response.answers.at("severity")).score;
  if (probability < 0 || probability > 1 || score < 0 || score > 3) throw std::runtime_error("Invalid bounds");
  // Apply thresholds and custom score scales in application code.
  std::cout << stringify(object({{"page_on_call", probability >= 0.9}, {"severity_1_to_5", 1 + 4 * score / 3}, {"response", response.to_value()}})) << "\n";
}
